tests/components/broadlink/test_device.py

```diff
@@ -0,0 +1,389 @@
+"""Tests for Broadlink devices."""
+import broadlink.exceptions as blke
+
+from homeassistant.components.broadlink.const import DOMAIN
+from homeassistant.components.broadlink.device import get_domains
+from homeassistant.config_entries import (
+    ENTRY_STATE_LOADED,
+    ENTRY_STATE_NOT_LOADED,
+    ENTRY_STATE_SETUP_ERROR,
+    ENTRY_STATE_SETUP_RETRY,
+)
+from homeassistant.helpers.entity_registry import async_entries_for_device
+
+from . import get_device
+
+from tests.async_mock import patch
+from tests.common import mock_device_registry, mock_registry
+
+
+async def test_device_setup(hass):
+    """Test a successful setup."""
+    device = get_device("Office")
+    mock_api = device.get_mock_api()
+    mock_entry = device.get_mock_entry()
+    mock_entry.add_to_hass(hass)
+
+    with patch("broadlink.gendevice", return_value=mock_api), patch.object(
+        hass.config_entries, "async_forward_entry_setup"
+    ) as mock_forward, patch.object(
+        hass.config_entries.flow, "async_init"
+    ) as mock_init:
+        await hass.config_entries.async_setup(mock_entry.entry_id)
+
+    assert mock_entry.state == ENTRY_STATE_LOADED
+    assert mock_api.auth.call_count == 1
+    assert mock_api.get_fwversion.call_count == 1
+    forward_entries = {c[1][1] for c in mock_forward.mock_calls}
+    domains = get_domains(mock_api.type)
+    assert mock_forward.call_count == len(domains)
+    assert forward_entries == domains
+    assert mock_init.call_count == 0
+
+
+async def test_device_setup_authentication_error(hass):
+    """Test we handle an authentication error."""
+    device = get_device("Living Room")
+    mock_api = device.get_mock_api()
+    mock_api.auth.side_effect = blke.AuthenticationError()
+    mock_entry = device.get_mock_entry()
+    mock_entry.add_to_hass(hass)
+
+    with patch("broadlink.gendevice", return_value=mock_api), patch.object(
+        hass.config_entries, "async_forward_entry_setup"
+    ) as mock_forward, patch.object(
+        hass.config_entries.flow, "async_init"
+    ) as mock_init:
+        await hass.config_entries.async_setup(mock_entry.entry_id)
+
+    assert mock_entry.state == ENTRY_STATE_SETUP_ERROR
+    assert mock_api.auth.call_count == 1
+    assert mock_forward.call_count == 0
+    assert mock_init.call_count == 1
+    assert mock_init.mock_calls[0][2]["context"]["source"] == "reauth"
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6042836)